Fix warnings reported by MSVC (`-W2`) and clang-cl (`-W4`)

MisterDA opened the pull request:
This PR fixes warnings raised by MSVC at level 2, and clang-cl at level 4 (minus unused parameters and sign-compare). It should be rather uneventful. I think it's sane to enable _some_ warnings rather than none at all (the current), but C compilers (and sometimes programmers too) have a tendency of being overzealous when they report issues.

MSVC really likes to warn when an assignment of two different ranges, like float to int, is made without an explicit cast. If the warnings [C4244](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4244?view=msvc-170) and [C4146](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4146?view=msvc-170) are too annoying, they could also be disabled.

Clang-cl `-Wall` is equivalent to clang `-Weverything`, which is *really* too noisy, and enables some C++ warnings. Levels above 2 for MSVC are also noisy, and some warnings are even checking for C++ things in C. Someone with spare time might revisit this later…

MisterDA committed:
Remove unreferenced local variables

MisterDA committed:
Work around _wmkdir not taking a permissions parameter

Avoids an unused variable warning for the perm value on Windows in:

    int perm = Int_val(vperm);
    ret = mkdir_os(path, perm);

MisterDA committed:
sak: remove unused macros, fix symbols visibility

MisterDA committed:
Fix MSVC warning C4244 converting from uintnat to double

> runtime/gc_ctrl.c: warning C4244: 'initializing': conversion from 'uintnat' to 'double', possible loss of data
> runtime/major_gc.c: warning C4244: 'initializing': conversion from 'uintnat' to 'double', possible loss of data
> runtime/major_gc.c: warning C4244: 'initializing': conversion from 'intnat' to 'double', possible loss of data
> runtime/major_gc.c: warning C4244: 'initializing': conversion from 'intnat' to 'double', possible loss of data

MisterDA committed:
Fix MSVC warning C4146 applying unary minus operator to unsigned

> runtime/ints.c: warning C4146: unary minus operator applied to unsigned type, result still unsigned
> runtime/backtrace_nat.c: warning C4146: unary minus operator applied to unsigned type, result still unsigned
> runtime/backtrace_nat.c: warning C4146: unary minus operator applied to unsigned type, result still unsigned

MisterDA committed:
Fix MSVC warning C4244 converting from int32_t to float

> runtime/memprof.c: warning C4244: 'initializing': conversion from 'int32_t' to 'float', possible loss of data

MisterDA committed:
Fix MSVC warning C4244 converting from double to uintnat

> runtime/major_gc.c: warning C4244: '=': conversion from 'double' to 'uintnat', possible loss of data
> runtime/major_gc.c: warning C4244: '=': conversion from 'double' to 'uintnat', possible loss of data

Co-authored-by: Xavier Leroy <xavier.leroy@college-de-france.fr>

MisterDA committed:
Fix MSVC warning C4146 applying unary minus operator to unsigned

> unixsupport_win32.c: warning C4146: unary minus operator applied to unsigned type, result still unsigned
> runtime/signals.c: warning C4146: unary minus operator applied to unsigned type, result still unsigned

MisterDA committed:
Fix minor int/size_t conversion warning

MisterDA committed:
Fix MSVC warning C4244 converting from ULONGLONG to double

gettimeofday_win32.c: warning C4244: =: conversion from ULONGLONG to double, possible loss of data

MisterDA committed:
Fix MSVC warning C4244 converting from double to DWORD

> sleep_win32.c: warning C4244: function: conversion from double to DWORD, possible loss of data
> select_win32.c: warning C4244: =: conversion from double to DWORD, possible loss of data

MisterDA committed:
Prevent incomplete initialization warning of a static value

Static values are zero-initialized by default.
clang-cl would warn that the remaining fields are missing
initializers.

MisterDA committed:
Enable some warnings of MSVC and clang-cl

MisterDA committed:
Prevent sign conversion warnings when decrementing with atomic_fetch_add

This prevents warnings when -1 is converted to an unsigned type.

MisterDA committed:
Port CAMLunused* to C23/C++17

This makes CAMLunused macros work with clang-cl, which doesn't define
__GNUC__ and doesn't support this MSVC compiler pragma.

MisterDA committed:
Remove workarounds for old versions of MSVC

No need to test for versions of MSVC before 1939, they don't support
C11 atomics.

MSVC now supports the inline C keyword.

No need for __assume(0) to mark unreachable code paths as ExitProcess
is already __declspec(noreturn).

nojb reviewed (commented):
Thanks! A chunk of the diff is dead code removal and clearly correct. Some of the other changes involve introducing explicit type casts; it would be good to have a second opinion because this can sometimes be tricky. Otherwise, I am happy to approve this PR.

@dustanddreams maybe you can take a look? Thanks!

dustanddreams commented on runtime/alloc.c:
```diff
@@ -78,13 +78,13 @@ CAMLexport value caml_alloc_shr_check_gc (mlsize_t wosize, tag_t tag)
+#define Enter_gc_preserve_vals(dom_st, wosize) do {                     \
+    CAMLparam0();                                                       \
+    CAMLlocalN(vals_copy, (wosize));                                    \
+    for (mlsize_t j = 0; j < (wosize); j++) vals_copy[j] = vals[j];     \
```
Why did you had to introduce a new local for the loop counter? `Enter_gc_preserve_vals` is only used from the `Do_alloc_small` macro next to it, which declares `mlsize_t i`.

Is there a warning about the use of `i` which is neither local to nor an argument of the macro? If so, such a too-pendantic warning ought to be disabled.

MisterDA replied:
`caml_alloc_9` defines parameters from `a` to `i`:
```c
CAMLexport value caml_alloc_9 (tag_t tag, value a, value b, value c, value d,
                               value e, value f, value g, value h, value i)
{
  Do_alloc_small(9, tag, a, b, c, d, e, f, g, h, i);
}
```
then `Do_alloc_small` expands to 
```c
#define Do_alloc_small(wosize, tag, ...)                \
{                                                       \
  Caml_check_caml_state();                              \
  value v;                                              \
  value vals[wosize] = {__VA_ARGS__};                   \
  mlsize_t i;                                           \
  CAMLassert ((tag) < 256);                             \
  // ...
```

the local `i` shadows the `i` parameter, thus the warning. Also, the `Enter_gc_preserve_vals` relies on the outer block declaring `i`, which I don't find particularly nice.
My intent was to kill two birds with one stone: first remove the shadowing warning by renaming the local `i` to `j`, then reduce the scope of the for loop iterators.

This was an incursion into MSVC level 4 warnings, so I'll just withdraw this commit.

dustanddreams commented on runtime/alloc.c:
```diff
@@ -95,12 +95,11 @@ CAMLexport value caml_alloc_shr_check_gc (mlsize_t wosize, tag_t tag)
   Alloc_small(v, wosize, tag, Enter_gc_preserve_vals);  \
-  for (i = 0; i < (wosize); i++) {                      \
-    Field(v, i) = vals[i];                              \
+  for (mlsize_t j = 0; j < (wosize); j++) {             \
```
Almost the same question here. Why did you have to rename the variable and reduce its scope? Did you get a shadowing warning? If so, wouldn't simply rename from `i` to something else in both macros would be good enough?

MisterDA replied:
See the answer above. I'll withdraw the commit.

dustanddreams commented on runtime/ints.c:
```diff
@@ -604,7 +604,7 @@ CAMLprim value caml_int64_of_string(value s)
     }
   }
-  if (sign < 0) res = - res;
+  if (sign < 0) res = ~res + 1;
```
Not really fond of this. Wouldn't `res = (uint64_t)-res` be good enough to silence those warnings?

xavierleroy replied:
I agree with @dustanddreams 's comment.  Writing `~x + 1` instead of `-x` is correct but really hard to read.

MisterDA replied:
`res` is already `uint64_t`. Not sure how `(uint64_t)-(int64_t)res` really behaves.
`res = UINT64_MAX - res + 1` isn't an improvement.
What if we went with a comment? `res = ~res + 1 /* res = -res */`

dustanddreams replied:
If `res` is already `int64_t`, then `(uint64_t)-res` will work nicely, since its value has already been bounds checked above so we know `0 <= res <= INT64_MAX`.

MisterDA replied:
I think you've misread, `res` is unsigned (`uint64_t`). The compiler warning is probably overzealous, but the compiler doesn't like when an unary minus is applied to an unsigned integer.

dustanddreams replied:
Ah, so two casts. Ugly but still more human-friendly than bitmasking.

MisterDA replied:
Turns out a single cast is sufficient.

xavierleroy commented on runtime/major_gc.c:
```diff
@@ -677,10 +677,10 @@ static void update_major_slice_work(intnat howmuch,
 
   total_cycle_work =
-    heap_sweep_words + (heap_words * 100 / (100 + caml_percent_free));
+    heap_sweep_words + (uintnat)(heap_words * 100 / (100 + caml_percent_free));
```
The original code is strange: why is `heap_words` a double-precision FP number when it obviously always holds integer values?  I'd rather declare `heap_words` as `uintnat` and fix the only place where it needs to be converted to FP:
```
  total_cycle_work =
    heap_sweep_words + (uintnat) ((double) heap_words * 100.0 / (100.0 + caml_percent_free));
```

MisterDA replied:
Done, thanks.

xavierleroy commented on runtime/caml/platform.h:
```diff
@@ -67,8 +67,8 @@ Caml_inline void cpu_relax(void) {
-Caml_inline uintnat atomic_fetch_add_verify_ge0(atomic_uintnat* p, uintnat v) {
-  uintnat result = atomic_fetch_add(p,v);
+Caml_inline uintnat atomic_fetch_sub_verify_ge0(atomic_uintnat* p, uintnat v) {
+  uintnat result = atomic_fetch_sub(p,v);
```
I'm used to "fetch and add", so "fetch and sub" feels a bit strange.  Why not declare the second parameter as `intnat v` and keep using "fetch and add -1" ?  This way the rest of the code doesn't need to be changed.

MisterDA replied:
It wouldn't work for direct calls to `atomic_fetch_sub`. There's just one of them, I'll keep it. Your idea works for the rest, thanks.

xavierleroy commented on runtime/signals.c:
```diff
@@ -505,16 +505,15 @@ static const int posix_signals[] = {
 {
-  int i;
-  for (i = 0; i < sizeof(posix_signals) / sizeof(int); i++)
+  for (int i = 0; i < (int)(sizeof(posix_signals) / sizeof(int)); i++)
```
Is the `(int)` cast necessary?  Why change the original code?

MisterDA replied:
I wanted to avoid a clang-cl warning, but I ended up disabling this warning (too much to "fix"). I also thought this change had a nice symmetry with the one above. 
```
runtime/signals.c(516,21): error: comparison of integers of different signs: 'int' and 'unsigned long long' [-Werror,-Wsign-compare]
  516 |   for (int i = 0; i < sizeof(posix_signals) / sizeof(int); i++)
      |                   ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

xavierleroy reviewed (commented):
I reviewed this PR a while ago but apparently forgot to submit the review.  I'm globally in favor of this PR, but there are a few places where I think a better fix is warranted, see below.

MisterDA commented:
I think I've addressed all the comments, many thanks for the reviews.

MisterDA committed:
Fix include style of system headers and of caml/ headers in tests

shindere commented:
@dustanddreams are you happy with the PR in its current state?

If so, I'm happy to approve on your behalf.

MisterDA commented:
It's conflicting with trunk now, I need to rebase.

shindere commented:
Antonin Décimo (2024/07/18 08:26 -0700):
> It's conflicting with trunk now, I need to rebase.

Oops, sorry, too quick.

MisterDA committed:
Fix MSVC warning C4244 converting from intnat to double

> runtime/major_gc.c(1353): warning C4244: 'initializing': conversion from 'intnat' to 'double', possible loss of data

MisterDA commented:
Rebased!

shindere commented:
Thanks. Happy to approve on somebody's behalf if necessary.

dustanddreams reviewed (approved):
🚀

MisterDA commented:
Rebased to fix conflicts.

shindere commented:
Thanks. Approving this on behalf of @dustanddreams.
